ThunkLibs/libwayland-client/Host.cpp

```diff
@@ -21,13 +22,281 @@ tags: thunklibs|wayland-client
+#ifdef IS_32BIT_THUNK
+  using type = uint32_t;
+#else
+  using type = wl_argument;
+#endif
+  type data;
+
+  guest_layout& operator=(const wl_argument from) {
+#ifdef IS_32BIT_THUNK
+    data = from.u;
+#else
+    data = from;
+#endif
+    return *this;
+  }
 };
 
+#include "thunkgen_host_libwayland-client.inl"
+
+// Maps guest interface to host_interfaces
+static std::unordered_map<guest_layout<const wl_interface>*, wl_interface*> guest_to_host_interface;
+
+static wl_interface* get_proxy_interface(wl_proxy* proxy) {
+  // wl_proxy is a private struct, but its first member is the wl_interface pointer
+  return *reinterpret_cast<wl_interface**>(proxy);
+}
+
+static void assert_is_valid_host_interface(const wl_interface* interface) {
+  if ((uint32_t)interface->method_count >= 0x1000 || (uint32_t)interface->event_count >= 0x1000) {
+    fprintf(stderr, "ERROR: Expected %p to be a host wl_interface, but it's not\n", interface);
+    std::abort();
+  }
+}
+
+static void assert_is_valid_guest_interface(guest_layout<const wl_interface*> guest_interface) {
+#ifdef IS_32BIT_THUNK
+  const wl_interface* as_host_interface = (const wl_interface*)guest_interface.force_get_host_pointer();
+  if ((uint32_t)as_host_interface->method_count < 0x1000 && (uint32_t)as_host_interface->event_count < 0x1000) {
+    fprintf(stderr, "ERROR: Expected %p to be a guest wl_interface, but it's not\n", guest_interface.force_get_host_pointer());
+    std::abort();
+  }
+#endif
+}
+
+#ifdef IS_32BIT_THUNK
+static void repack_guest_wl_interface_to_host(guest_layout<const wl_interface*> guest_interface_ptr, wl_interface* host_interface) {
+  auto& guest_interface = *guest_interface_ptr.get_pointer();
+  static_assert(sizeof(guest_interface) == 24);
+
+  *host_interface = host_layout<wl_interface> { guest_interface }.data;
+  fex_apply_custom_repacking_entry(reinterpret_cast<host_layout<wl_interface>&>(*host_interface), guest_interface);
+}
+
+// Maps guest interface pointers to host pointers
+static const wl_interface* lookup_wl_interface(guest_layout<const wl_interface*> interface) {
+  // Used e.g. for wl_shm_pool_destroy
+  if (interface.force_get_host_pointer() == nullptr) {
+    return nullptr;
+  }
+
+  auto [host_interface_it, inserted] = guest_to_host_interface.emplace(interface.get_pointer(), nullptr);
+  if (!inserted) {
+    assert_is_valid_host_interface(host_interface_it->second);
+    return host_interface_it->second;
+  }
+
+  assert_is_valid_guest_interface(interface);
+
+  fprintf(stderr, "Unknown wayland interface %p, adding to registry\n", interface.get_pointer());
+
+  host_interface_it->second = new wl_interface;
+  wl_interface* host_interface = host_interface_it->second;
+  repack_guest_wl_interface_to_host(interface, host_interface);
+  return host_interface_it->second;
+}
+
+void fex_custom_repack_entry(host_layout<wl_interface>& into, guest_layout<wl_interface> const& from) {
+  // TODO: Manage lifetime of this memory?
```

Comment:
This isn't actually needed. `wl_interface`s correspond to global symbols on the guest side, so their lifetime is bound to program execution.